Add native types to CategoryBundle

Prokyonn opened the pull request:
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | yes
| Deprecations? | no
| Related issues/PRs | depends on https://github.com/sulu/sulu/pull/8405
| License | MIT

#### What's in this PR?

Adds native types to entities and their methods.

#### Why?

Better developer experience.

Prokyonn committed:
Add native types to CategoryBundle entities and interfaces

alexander-schranz commented on src/Sulu/Bundle/CategoryBundle/Entity/Category.php:
```diff
@@ -82,125 +61,125 @@ public function __construct()
     }
 
-    public function findTranslationByLocale($locale)
+    public function findTranslationByLocale(?string $locale): CategoryTranslationInterface|false
```
@Prokyonn should we convert this to null via `->first() ?: null` so null safe operator `?->` can used on it?

alexander-schranz commented on src/Sulu/Bundle/CategoryBundle/Entity/Category.php:
```diff
@@ -209,19 +188,19 @@ function(CategoryTranslationInterface $translation) use ($locale) {
     }
 
-    public function removeChild(CategoryInterface $child)
+    public function removeChild(CategoryInterface $child): void
```
would `return $this` (`: static`) in all methods which currently may are `void`

alexander-schranz commented on src/Sulu/Bundle/CategoryBundle/Entity/CategoryMeta.php:
```diff
@@ -16,90 +16,72 @@
  */
 class CategoryMeta implements CategoryMetaInterface
```
what does this entity 🤔  ? is it editable by the UI?

mamazu commented on src/Sulu/Bundle/TagBundle/Entity/Tag.php:
```diff
@@ -90,7 +84,7 @@ public function setChanged(\DateTimeImmutable $changed): self
-    public function __toString()
+    public function __toString(): string
     {
         return (string) $this->getName();
```
Since getName is typed to a string, we don't need to cast here.

alexander-schranz replied:
we need set atleast a default for `$this->name` as `__toString()` is always called by Doctrine if exception happening and then instead of showing that exception it shows null to string error: https://github.com/sulu/sulu/pull/7478

mamazu replied:
Then maybe something like `return $this->name ?? 'unnamed';`?

alexander-schranz replied:
would keep as before empty string, doctrine fallbacks to a object hash I think then like when no __toString exists which is better in the trace.

alexander-schranz commented:
@Prokyonn did discuss with @chirimoya that we can remove the unused `CategoryMeta` and `metas` from the category entity: https://github.com/sulu/sulu/blob/3.0/src/Sulu/Bundle/CategoryBundle/Entity/CategoryMeta.php

Prokyonn committed:
Remove CategoryMeta

Prokyonn committed:
Remove nullable ids

Prokyonn commented on src/Sulu/Bundle/CategoryBundle/Category/KeywordManager.php:
```diff
@@ -37,7 +37,7 @@ public function __construct(
     }
 
-    public function save(KeywordInterface $keyword, CategoryInterface $category, $force = null)
+    public function save(KeywordInterface $keyword, CategoryInterface $category, $force = null, bool $isPersisted = false)
```
@wachterjohannes remove isPersisted again (we use `isNew` from the entity directly)

Prokyonn commented on src/Sulu/Bundle/CategoryBundle/Category/KeywordManager.php:
```diff
@@ -48,7 +48,7 @@ public function save(KeywordInterface $keyword, CategoryInterface $category, $fo
         }
 
-        if (null !== $keyword->getId()
+        if ($isPersisted
```
replace with $keyword->isNew

Prokyonn commented on src/Sulu/Bundle/CategoryBundle/Category/KeywordManager.php:
```diff
@@ -63,7 +63,7 @@ public function save(KeywordInterface $keyword, CategoryInterface $category, $fo
         /** @var CategoryTranslationInterface $categoryTranslation */
         foreach ($keyword->getCategoryTranslations() as $categoryTranslation) {
-            $event = $keyword->getId()
+            $event = $isPersisted
```
replace with $keyword-isNew

Prokyonn commented on src/Sulu/Bundle/CategoryBundle/Category/KeywordManagerInterface.php:
```diff
@@ -32,7 +32,7 @@ interface KeywordManagerInterface
      * @return KeywordInterface
      */
-    public function save(KeywordInterface $keyword, CategoryInterface $category, $force = null);
+    public function save(KeywordInterface $keyword, CategoryInterface $category, $force = null, bool $isPersisted = false);
```
@wachterjohannes remove isPersisted again (we use `isNew` from the entity directly)

wachterjohannes committed:
Replace isPersisted parameter with isNew() check in KeywordManager

- Remove bool $isPersisted parameter from KeywordManager::save() method
- Use $keyword->isNew() to determine if keyword is new or persisted
- Update KeywordController to not pass isPersisted parameter
- Fix unit tests to properly mock isNew() for both keyword and otherKeyword
- Update event dispatching in tests based on keyword state

This addresses PR review comments to use entity's isNew() method
directly instead of passing state as a parameter.

wachterjohannes committed:
Fix CategoryKeywordRemovedEvent to store category ID directly

- Add categoryId parameter to CategoryKeywordRemovedEvent constructor
- Use stored categoryId instead of calling category->getId()
- Update KeywordManager to pass category ID when creating event

This fixes the error: 'Typed property Category::$id must not be
accessed before initialization' which occurred when deleting categories
with keywords. The issue was that the event tried to access getId() on
a category entity that was in a transitional state during deletion,
where the ID property was no longer accessible.

Following the pattern from CategoryRemovedEvent, we now capture the
category ID before the entity state changes and store it directly in
the event.
